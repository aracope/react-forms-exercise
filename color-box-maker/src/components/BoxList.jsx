import React, { useState } from "react";
import NewBoxForm from "./NewBoxForm";
import Box from "./Box";
import { v4 as uuid } from "uuid";

function BoxList() {
  const [boxes, setBoxes] = useState([]);

  const addBox = (boxData) => {
    setBoxes((boxes) => [...boxes, { ...boxData, id: uuid() }]);
  };

  const removeBox = (id) => {
    setBoxes((boxes) => boxes.filter((box) => box.id !== id));
  };

  return (
    <div>
      <NewBoxForm addBox={addBox} />
      <div className="box-container">
        {boxes.map((box) => (
          <Box
            key={box.id}
            id={box.id}
            width={box.width}
            height={box.height}
            backgroundColor={box.backgroundColor}
            removeBox={removeBox}
          />
        ))}
      </div>
    </div>
  );
}

export default BoxList;
