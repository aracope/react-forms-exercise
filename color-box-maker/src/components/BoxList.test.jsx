import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";



// Smoke
it("renders without crashing", () => {
  render(<BoxList />);
});

// Snapshot
it("matches snapshot", () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

// Add and remove box
it("can add and remove a box", () => {
  const { getByLabelText, getByText, queryByTestId } = render(<BoxList />);
  const widthInput = getByLabelText(/width/i);
  const heightInput = getByLabelText(/height/i);
  const colorInput = getByLabelText(/background color/i);

  fireEvent.change(widthInput, { target: { value: "100" } });
  fireEvent.change(heightInput, { target: { value: "100" } });
  fireEvent.change(colorInput, { target: { value: "red" } });
  fireEvent.click(getByText("Add Box"));

  const box = queryByTestId("box");
  expect(box).toBeInTheDocument();

  const removeBtn = getByText("X");
  fireEvent.click(removeBtn);

  expect(queryByTestId("box")).not.toBeInTheDocument();
});
