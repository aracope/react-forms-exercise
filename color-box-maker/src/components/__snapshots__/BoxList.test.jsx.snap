// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label>
        Width:
        <input
          name="width"
          type="number"
          value=""
        />
      </label>
      <label>
        Height:
        <input
          name="height"
          type="number"
          value=""
        />
      </label>
      <label>
        Background Color:
        <input
          name="backgroundColor"
          type="text"
          value=""
        />
      </label>
      <button>
        Add Box
      </button>
    </form>
    <div
      class="box-container"
    />
  </div>
</DocumentFragment>
`;
